Debug Zip Missing Panic Logs

**Describe the problem**

I collected a debug zip for a cluster that recently had nodes restart due to a panic. I was taking the debug zip in order to determine why a node was panicing. The stderr log containing the unexpected panic was not included in the debug zip. 

The debug zip logs contained this error that was generated by taking the debug zip. The error seems to suggest CRDB was unable to parse the stderr logs including the panic:

```
E230601 14:51:13.619603 86574 server/status.go:1310 ⋮ [T1,n1,status] 581 decoding on line 8: malformed log entry
(1) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/util/log.(*entryDecoderV2).Decode.func1
  | 	github.com/cockroachdb/cockroach/pkg/util/log/format_crdb_v2.go:528
  | runtime.gopanic
  | 	GOROOT/src/runtime/panic.go:884
  | [...repeated from below...]
Wraps: (2) decoding on line 8
Wraps: (3) attached stack trace
  -- stack trace:
  | github.com/cockroachdb/cockroach/pkg/util/log.(*entryDecoderV2).peekNextFragment
  | 	github.com/cockroachdb/cockroach/pkg/util/log/format_crdb_v2.go:615
  | github.com/cockroachdb/cockroach/pkg/util/log.(*entryDecoderV2).Decode
  | 	github.com/cockroachdb/cockroach/pkg/util/log/format_crdb_v2.go:548
  | github.com/cockroachdb/cockroach/pkg/server.(*statusServer).LogFile
  | 	github.com/cockroachdb/cockroach/pkg/server/status.go:1306
  | github.com/cockroachdb/cockroach/pkg/server/serverpb._Status_LogFile_Handler.func1
  | 	github.com/cockroachdb/cockroach/pkg/server/serverpb/bazel-out/k8-opt/bin/pkg/server/serverpb/serverpb_go_proto_/github.com/cockroachdb/cockroach/pkg/server/serverpb/status.pb.go:9040
  | github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.ServerInterceptor.func1
  | 	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:115
  | google.golang.org/grpc.chainUnaryInterceptors.func1.1
  | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1162
  | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3
  | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:276
  | google.golang.org/grpc.chainUnaryInterceptors.func1.1
  | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1165
  | github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor
  | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:105
  | google.golang.org/grpc.chainUnaryInterceptors.func1.1
  | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1165
  | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1
  | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:243
  | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr
  | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:322
  | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1
  | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:241
  | google.golang.org/grpc.chainUnaryInterceptors.func1.1
  | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1165
  | google.golang.org/grpc.chainUnaryInterceptors.func1
  | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1167
  | github.com/cockroachdb/cockroach/pkg/server/serverpb._Status_LogFile_Handler
  | 	github.com/cockroachdb/cockroach/pkg/server/serverpb/bazel-out/k8-opt/bin/pkg/server/serverpb/serverpb_go_proto_/github.com/cockroachdb/cockroach/pkg/server/serverpb/status.pb.go:9042
  | google.golang.org/grpc.(*Server).processUnaryRPC
  | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1340
  | google.golang.org/grpc.(*Server).handleStream
  | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1713
  | google.golang.org/grpc.(*Server).serveStreams.func1.2
  | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:965
  | runtime.goexit
  | 	GOROOT/src/runtime/asm_amd64.s:1594
Wraps: (4) malformed log entry
Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.leafError
```

The content of the std error file was:

```
sh-4.4# cat cockroach-stderr.cockroachdb-0.root.2023-05-31T21_52_20Z.000030.log
I230531 21:52:20.499220 1 util/log/file_sync_buffer.go:238 ⋮ [config]   file created at: 2023/05/31 21:52:20
I230531 21:52:20.499241 1 util/log/file_sync_buffer.go:238 ⋮ [config]   running on machine: cockroachdb-0
I230531 21:52:20.499248 1 util/log/file_sync_buffer.go:238 ⋮ [config]   binary: CockroachDB CCL v23.1.2 (x86_64-pc-linux-gnu, built 2023/05/25 15:51:13, go1.19.4)
I230531 21:52:20.499255 1 util/log/file_sync_buffer.go:238 ⋮ [config]   arguments: [/cockroach/cockroach start --certs-dir /cockroach/cockroach-certs --listen-addr 0.0.0.0:26258 --http-addr 0.0.0.0:8080 --sql-addr 0.0.0.0:26257 --advertise-addr cockroachdb-0.cockroachdb.us-west-2.svc.cluster.local:26258 --advertise-sql-addr cockroachdb-0.cockroachdb.us-west-2.svc.cluster.local:26257 --join _grpc._tcp.cockroachdb.us-west-2.svc --experimental-dns-srv --cache 25% --max-sql-memory 25% --locality=region=aws-us-west-2,az=aws-us-west-2a,dns=cockroachdb-0.cockroachdb.us-west-2.svc.cluster.local --log-config-file /cockroach/config/logs.yaml --log-config-vars HOST_IP --external-io-dir=disabled --max-offset=100ms]
I230531 21:52:20.499284 1 util/log/file_sync_buffer.go:238 ⋮ [config]   log format (utf8=✓): crdb-v2
I230531 21:52:20.499288 1 util/log/file_sync_buffer.go:238 ⋮ [config]   line format: [IWEF]yymmdd hh:mm:ss.uuuuuu goid [chan@]file:line redactionmark \[tags\] [counter] msg
I230531 21:52:20.498414 1 util/log/flags.go:211  [-] 1  stderr capture started
panic: invalid tenant ID 0

goroutine 6603729 [running]:
github.com/cockroachdb/cockroach/pkg/roachpb.assertValid(0x5398fe0?)
        github.com/cockroachdb/cockroach/pkg/roachpb/pkg/roachpb/tenant.go:99 +0x59
github.com/cockroachdb/cockroach/pkg/roachpb.MustMakeTenantID(...)
        github.com/cockroachdb/cockroach/pkg/roachpb/pkg/roachpb/tenant.go:52
github.com/cockroachdb/cockroach/pkg/ccl/backupccl.backupShowerDefault.func1({0x7281e90, 0xc00644dd40}, {{0xc008bfee10, 0x2e0}, {0xc00c389130, 0x1, 0x1}, {0xc001812600, 0x1, 0x1}, ...})
        github.com/cockroachdb/cockroach/pkg/ccl/backupccl/show.go:967 +0x24eb
github.com/cockroachdb/cockroach/pkg/ccl/backupccl.manifestInfoReader.showBackup({{{0xc008961500?, 0x0?, 0x0?}, 0xc00c284120?}}, {0x7281e90?, 0xc00c284fc0?}, _, _, {{0xc008bfee10, 0x2e0}, ...}, ...)
        github.com/cockroachdb/cockroach/pkg/ccl/backupccl/show.go:120 +0x205
github.com/cockroachdb/cockroach/pkg/ccl/backupccl.showBackupPlanHook.func1({0x7281e90, 0xc00c284f00}, {0x0?, 0x0?, 0x0?}, 0xc00e8503c0?)
        github.com/cockroachdb/cockroach/pkg/ccl/backupccl/show.go:531 +0x1fc5
github.com/cockroachdb/cockroach/pkg/sql.(*hookFnNode).startExec.func1()
        github.com/cockroachdb/cockroach/pkg/sql/planhook.go:205 +0xa4
created by github.com/cockroachdb/cockroach/pkg/sql.(*hookFnNode).startExec
        github.com/cockroachdb/cockroach/pkg/sql/planhook.go:203 +0x1b2
```


Epic CRDB-32134

--- a/pkg/cli/zip_per_node.go
+++ b/pkg/cli/zip_per_node.go
@@ -460,15 +460,12 @@ func (zc *debugZipContext) collectPerNodeData(
 			// Log the list of errors that occurred during log entries request.
 			if len(entries.ParseErrors) > 0 {
 				sf.shout("%d parsing errors occurred:", len(entries.ParseErrors))
-				for _, err := range entries.ParseErrors {
-					sf.shout("%s", err)
-				}
 				parseErr := fmt.Errorf("%d errors occurred:\n%s", len(entries.ParseErrors), strings.Join(entries.ParseErrors, "\n"))
 				if err := zc.z.createError(sf, name, parseErr); err != nil {
 					return err
 				}
 			}
-			sf.progress("writing output: %s", name)
+			sf = logPrinter.start("writing output")
 			warnRedactLeak := false
 			if err := func() error {
 				// Use a closure so that the zipper is only locked once per

--- a/pkg/server/status.go
+++ b/pkg/server/status.go
@@ -1367,6 +1367,8 @@ func (s *statusServer) LogFile(
 			}
 			if errors.Is(err, log.ErrMalformedLogEntry) {
 				resp.ParseErrors = append(resp.ParseErrors, err.Error())
+				//Append log generated from malformed line.
+				resp.Entries = append(resp.Entries, entry)
 				// Proceed decoding next entry, as we want to retrieve as much logs
 				// as possible.
 				continue

--- a/pkg/server/storage_api/logfiles_test.go
+++ b/pkg/server/storage_api/logfiles_test.go
@@ -513,5 +513,5 @@ func TestStatusLogCorruptedEntry(t *testing.T) {
 	require.NoError(t, err)
 	require.NotEmpty(t, wrapper.Entries)
 	require.NotEmpty(t, wrapper.ParseErrors)
-	require.Equal(t, len(wrapper.ParseErrors), 1)
+	require.Equal(t, 2, len(wrapper.ParseErrors))
 }

--- a/pkg/util/log/format_crdb_v2.go
+++ b/pkg/util/log/format_crdb_v2.go
@@ -460,6 +460,7 @@ func (buf *buffer) maybeMultiLine(
 var startRedactionMarker = string(redact.StartMarker())
 var endRedactionMarker = string(redact.EndMarker())
 var markersStartWithMultiByteRune = startRedactionMarker[0] >= utf8.RuneSelf && endRedactionMarker[0] >= utf8.RuneSelf
+var defaultCrdbV2Format = "I000101 00:00:00.000000 0  :0 ⋮ [-] 0  %s"
 
 var (
 	entryREV2 = regexp.MustCompile(
@@ -495,13 +496,16 @@ var (
 const tenantDetailsTags = string(tenantIDLogTagKey) + string(tenantNameLogTagKey)
 
 type entryDecoderV2 struct {
-	lines           int // number of lines read from reader
-	reader          *bufio.Reader
-	nextFragment    entryDecoderV2Fragment
-	sensitiveEditor redactEditor
+	lines               int // number of lines read from reader
+	reader              *bufio.Reader
+	nextFragment        entryDecoderV2Fragment
+	sensitiveEditor     redactEditor
+	isMalformedFragment bool
 }
 
 // Decode decodes the next log entry into the provided protobuf message.
+// If we encounter any malformed line then we are updating log entry along with
+// MalformedLogEntry error
 func (d *entryDecoderV2) Decode(entry *logpb.Entry) (err error) {
 	defer func() {
 		if err != nil && !errors.Is(err, io.EOF) {
@@ -509,13 +513,22 @@ func (d *entryDecoderV2) Decode(entry *logpb.Entry) (err error) {
 		}
 	}()
 	frag, err := d.peekNextFragment()
+
 	if err != nil {
 		return err
 	}
-	d.popFragment()
-	if err = d.initEntryFromFirstLine(entry, frag); err != nil {
-		return err
+
+	// construct the log entry from malformed fragment and return it along
+	// with malformed log entry error
+	if d.isMalformedFragment {
+		d.popFragment()
+		d.initEntryFromFirstLine(entry, frag)
+		entry.Message = string(frag.getMsg())
+		return ErrMalformedLogEntry
 	}
+
+	d.popFragment()
+	d.initEntryFromFirstLine(entry, frag)
 
 	// Process the message.
 	var entryMsg bytes.Buffer
@@ -524,8 +537,8 @@ func (d *entryDecoderV2) Decode(entry *logpb.Entry) (err error) {
 	// While the entry has additional lines, collect the full message.
 	for {
 		frag, err = d.peekNextFragment()
-		if err != nil || !frag.isContinuation() {
-			// Ignore this error as it is relevant to the next line and we don't
+		if err != nil || d.isMalformedFragment || !frag.isContinuation() {
+			// Ignore the error or malformed fragment as it is relevant to the next line and we don't
 			// know if it is continuation line or not.
 			break
 		}
@@ -595,7 +608,10 @@ func (d *entryDecoderV2) peekNextFragment() (entryDecoderV2Fragment, error) {
 			if d.lines == 1 { // allow non-matching lines if we've never seen a line
 				continue
 			}
-			return nil, ErrMalformedLogEntry
+			// construct fragment based on malformed line in crdbV2 format with default values
+			d.nextFragment = entryREV2.FindSubmatch([]byte((fmt.Sprintf(defaultCrdbV2Format, nextLine))))
+			d.isMalformedFragment = true
+			return d.nextFragment, nil
 		}
 		d.nextFragment = m
 	}
@@ -607,11 +623,10 @@ func (d *entryDecoderV2) popFragment() {
 		panic(errors.AssertionFailedf("cannot pop unpopulated fragment"))
 	}
 	d.nextFragment = nil
+	d.isMalformedFragment = false
 }
 
-func (d *entryDecoderV2) initEntryFromFirstLine(
-	entry *logpb.Entry, m entryDecoderV2Fragment,
-) (err error) {
+func (d *entryDecoderV2) initEntryFromFirstLine(entry *logpb.Entry, m entryDecoderV2Fragment) {
 	// Erase all the fields, to be sure.
 	tenantID, tenantName := m.getTenantDetails()
 	*entry = logpb.Entry{
@@ -631,7 +646,6 @@ func (d *entryDecoderV2) initEntryFromFirstLine(
 		entry.StructuredStart = 0
 		entry.StructuredEnd = uint32(len(m.getMsg()))
 	}
-	return nil
 }
 
 // entryDecoderV2Fragment is a line which is part of a v2 log entry.

--- a/pkg/util/log/format_crdb_v2_test.go
+++ b/pkg/util/log/format_crdb_v2_test.go
@@ -268,6 +268,7 @@ func TestCrdbV2Decode(t *testing.T) {
 							break
 						}
 						if errors.Is(err, ErrMalformedLogEntry) {
+							fmt.Fprintf(&out, "malformed entry:%# v\n", pretty.Formatter(e))
 							continue
 						}
 						td.Fatalf(t, "error while decoding: %v", err)

--- a/pkg/util/log/testdata/parse
+++ b/pkg/util/log/testdata/parse
@@ -279,14 +279,48 @@ subtest corrupted_log
 
 # Check that entries after corrupted line is parsed.
 log
-.
 I210116 17:49:17.073282-020000 14 server/node.go:464 ⋮ [-] 23  started with engine type ‹2›
-.
-.
-.
+panic: invalid tenant ID 0
+
+goroutine 6603729 [running]:
 I210116 19:49:17.073282 14 server/node.go:464 ⋮ [-] 23  last line
 ----
 logpb.Entry{Severity:1, Time:1610826557073282000, Goroutine:14, File:"server/node.go", Line:464, Message:"started with engine type ‹2›", Tags:"", Counter:0x17, Redactable:true, Channel:0, StructuredEnd:0x0, StructuredStart:0x0, StackTraceStart:0x0, TenantID:"1", TenantName:""}
+malformed entry:logpb.Entry{Severity:1, Time:946684800000000000, Goroutine:0, File:" ", Line:0, Message:"panic: invalid tenant ID 0", Tags:"", Counter:0x0, Redactable:true, Channel:0, StructuredEnd:0x0, StructuredStart:0x0, StackTraceStart:0x0, TenantID:"1", TenantName:""}
+malformed entry:logpb.Entry{Severity:1, Time:946684800000000000, Goroutine:0, File:" ", Line:0, Message:"", Tags:"", Counter:0x0, Redactable:true, Channel:0, StructuredEnd:0x0, StructuredStart:0x0, StackTraceStart:0x0, TenantID:"1", TenantName:""}
+malformed entry:logpb.Entry{Severity:1, Time:946684800000000000, Goroutine:0, File:" ", Line:0, Message:"goroutine 6603729 [running]:", Tags:"", Counter:0x0, Redactable:true, Channel:0, StructuredEnd:0x0, StructuredStart:0x0, StackTraceStart:0x0, TenantID:"1", TenantName:""}
+logpb.Entry{Severity:1, Time:1610826557073282000, Goroutine:14, File:"server/node.go", Line:464, Message:"last line", Tags:"", Counter:0x17, Redactable:true, Channel:0, StructuredEnd:0x0, StructuredStart:0x0, StackTraceStart:0x0, TenantID:"1", TenantName:""}
+
+
+subtest end
+
+subtest corrupted_log_at_end
+
+# Check that entries after corrupted at end are getting parsed
+log
+I210116 17:49:17.073282-020000 14 server/node.go:464 ⋮ [-] 23  started with engine type ‹2›
+panic: invalid tenant ID 0
+
+goroutine 6603729 [running]:
+----
+logpb.Entry{Severity:1, Time:1610826557073282000, Goroutine:14, File:"server/node.go", Line:464, Message:"started with engine type ‹2›", Tags:"", Counter:0x17, Redactable:true, Channel:0, StructuredEnd:0x0, StructuredStart:0x0, StackTraceStart:0x0, TenantID:"1", TenantName:""}
+malformed entry:logpb.Entry{Severity:1, Time:946684800000000000, Goroutine:0, File:" ", Line:0, Message:"panic: invalid tenant ID 0", Tags:"", Counter:0x0, Redactable:true, Channel:0, StructuredEnd:0x0, StructuredStart:0x0, StackTraceStart:0x0, TenantID:"1", TenantName:""}
+malformed entry:logpb.Entry{Severity:1, Time:946684800000000000, Goroutine:0, File:" ", Line:0, Message:"", Tags:"", Counter:0x0, Redactable:true, Channel:0, StructuredEnd:0x0, StructuredStart:0x0, StackTraceStart:0x0, TenantID:"1", TenantName:""}
+malformed entry:logpb.Entry{Severity:1, Time:946684800000000000, Goroutine:0, File:" ", Line:0, Message:"goroutine 6603729 [running]:", Tags:"", Counter:0x0, Redactable:true, Channel:0, StructuredEnd:0x0, StructuredStart:0x0, StackTraceStart:0x0, TenantID:"1", TenantName:""}
+
+
+subtest end
+
+
+subtest corrupted_log_at_start
+
+# Check that entries after corrupted at start are getting parsed
+log
+goroutine 6603729 [running]:
+panic: invalid tenant ID 0
+I210116 19:49:17.073282 14 server/node.go:464 ⋮ [-] 23  last line
+----
+malformed entry:logpb.Entry{Severity:1, Time:946684800000000000, Goroutine:0, File:" ", Line:0, Message:"panic: invalid tenant ID 0", Tags:"", Counter:0x0, Redactable:true, Channel:0, StructuredEnd:0x0, StructuredStart:0x0, StackTraceStart:0x0, TenantID:"1", TenantName:""}
 logpb.Entry{Severity:1, Time:1610826557073282000, Goroutine:14, File:"server/node.go", Line:464, Message:"last line", Tags:"", Counter:0x17, Redactable:true, Channel:0, StructuredEnd:0x0, StructuredStart:0x0, StackTraceStart:0x0, TenantID:"1", TenantName:""}
 
 

